Fixed #24301 - Added general and statistics aggregate functions for PostgreSQL

coldmind opened the pull request:
https://code.djangoproject.com/ticket/24301

Feature is implemented (for general and statistics functions.)

charettes commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -0,0 +1,256 @@
+    fixtures = ["aggregation_general.json"]
+
+    def test_array_agg_charfield(self):
+        values = GeneralTestModel.objects.all().aggregate(ArrayAgg('char_field'))
```
All the `all().aggregate()` calls can be replaced by `aggregate()` calls.

charettes commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,120 @@
+            raise TypeError('Both X and Y must be provided.')
+        if (not isinstance(x, (six.text_type, six.string_types)) or
+                not isinstance(y, (six.text_type, six.string_types))):
+            raise ValueError('X and Y must be a string.')
```
Why can't `x` or `y` be specified as `int`?

coldmind replied:
Later, in `parse_expressions`, there is a `.isdigit()` check.
I think it is better to allow only strings than first check if it is a number and then cast it to string.

charettes replied:
Why do you need to cast it as a string? Shouldn't this parameter be casted to the correct type by the database backend (`psycopg2` in this case)?

coldmind replied:
Before it will go to database, we need to decide what expression it will be (`F()` or `Value()`).
Because if we will pass an integer to `F()`, it will raise an exception that it is not a field name.
But we need to allow integer to be passed to final SQL-query, because postgres allow integer in statistic functions.

charettes replied:
Ok I understand that you have to determine if it's either a field reference or an explicit value but I think you should do it the other way around. That is accept both `six.text_type` and `six.integer_types` and consider `six.text_type` as field references (`F()`) and everything else as an explicit value (`Value()`).

coldmind replied:
OK, it sounds reasonable, will do.

charettes commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,120 @@
+            sql_parts.append(arg_sql)
+            params.extend(arg_params)
+        mapping = {'function': self.function, 'y': sql_parts[1], 'x': sql_parts[0]}
+        return self.template % mapping, params
```
Shouldn't `x` and `y` be returned as parameters instead?

coldmind replied:
No, because, for example, if we will pass an integer to, e.g., `x`, it will be placed to SQL after `self.template % mapping` as `%s`, and this integer value will be placed to `params`.
Later, when sql will be executed, it will pass integer param to `%s`, so everything will work as expected if I understand correctly.
You can check it with debugger.
I wrote somewhat "complex" test, and result value of calculated sql is correct.

charettes replied:
Ok I get it, `compiler.compile(Value(5))` would return something like `'%d', (5,)`.

jarshwah replied:
This as_sql method can be deleted if you remove the custom template and implement get/set_source_expressions as above.

charettes commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,120 @@
+        if (not isinstance(x, (six.text_type, six.string_types)) or
+                not isinstance(y, (six.text_type, six.string_types))):
+            raise ValueError('X and Y must be a string.')
+        super(StatFunc, self).__init__(output_field=output_field)
```
Can't you just pass and `x` and `y` to `__init__` so they endup in `self.extra` and you don't have to override `as_sql`?

charettes commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,120 @@
+        # this case.
+        return [
+            Value(arg) if arg.isdigit() else F(arg) for arg in expressions
+        ]
```
I wonder if we shouldn't simply change `Func._parse_expressions` to do the following intead:

```python
def _parse_expressions(self, *expressions):
        return [
            arg if hasattr(arg, 'resolve_expression') else (
               F(arg) if isinstance(arg, six.string_types) else Value(arg)
            ) for arg in expressions
        ]
```

What do you think @jarshwah?

coldmind replied:
I wrote `parse_expressions` (without underscore) because it didn't wanted to be overriden (some issue with multiple inheritance), I will try to override it again.

jarshwah replied:
@coldmind I think @charettes was suggesting change the original `_parse_expressions` to differentiate between strings and numbers, and I think that's a great idea.

Rather than define the method above on `StatFunc`, we should change the method directly on `Func` which will then propagate to StatFunc through inheritance.

It may be a little magical, but the rules can be defined quite clearly. "string is a field ref", "resolve_expression passed verbatim", everything else is a "Value(value)".

I like this enough to write a patch today for inclusion in 1.8beta - and update all the applicable tests. Will land it today hopefully, and then we can rebase this PR here.

MarkusH commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -0,0 +1,47 @@
+from django.db.models.aggregates import Aggregate
+
+__all_ = [
```
Missing duplicate trailing underscore: `__all__`

MarkusH commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,118 @@
+from django.db.models.expressions import Value
+from django.utils import six
+
+__all_ = [
```
As above

MarkusH commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,212 @@
+PostgreSQL specific aggregation functions
+=========================================
+
+All of these functions are available from the ``django.contrib.postgres.aggregates``
```
You should put a `.. module::` directive here. See https://github.com/django/django/blob/master/docs/ref/models/meta.txt#L5-L6 for an example.

MarkusH commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,212 @@
+PostgreSQL specific aggregation functions
```
The main headline has `===` above and below.

jarshwah commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,118 @@
+        # we need to have kinda "static" alias for this case.
+        x = self._num_expression_alias if not isinstance(self.x, (six.string_types, six.text_type)) else self.x
+        y = self._num_expression_alias if not isinstance(self.y, (six.string_types, six.text_type)) else self.y
+        return '%s_%s__%s' % (y, x, self.name.lower())
```
I think this would probably cause more confusion than it's worth. I'd be in favour of not supporting default_alias for new aggregates, and requiring users to supply their own alias. I'm biased against default_alias though, so I won't veto. Interested in what others thing.

coldmind replied:
It will require users to do specify alias each time or to subclass aggregate function and set own subclass.
Each of this is not good.
If user will want to custom alias, he can always do this.

jarshwah replied:
```
.aggregate(regr=RegrSYY('col_a', 'col_b'))
# vs
.aggregate(RegrSYY('col_a', 'col_b')) -> col_a_col_b_regrsyy
```

It's a long and unwieldy default alias. You also have to be careful about the column name length in this situation. Oracle only supports names up to 30 chars, which means both column names need to be 10 chars or less to account for the extra bits of the underscores and aggregate name. I don't see the value.

coldmind replied:
Well, these are postgres-specific aggregates, so Oracle can sleep peacefully :)

I think it will be better to have the same behaviour for new aggregates as in old (with default alias).
I agree that static function aliases looks not good, but they are self-describing (you can see where is X and where is Y).
Maybe we should give an ability users to override default alias in the way from your example: `.aggregate(regr=RegrSYY('col_a', 'col_b'))`

jarshwah commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,118 @@
+        super(StatFunc, self).__init__(y=y, x=x, output_field=output_field)
+        self.x = x
+        self.y = y
+        self.source_expressions = self._parse_expressions(self.x, self.y)
```
The problem with this line is that further down you're referencing `self.x` and `self.y` directly. If another part of the code calls `set_source_expressions`, you're going to be using old values.

You should override get/set_source_expressions to read from and assign to self.x, self.y.

jarshwah commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,118 @@
+
+
+class StatFunc(Aggregate):
+    template = "%(function)s(%(y)s, %(x)s)"
```
You don't need to change the template here as long as get/set_source_expressions reads/writes x/y in the right order.

```
    def get_source_expressions(self):
        return self.y, self.x

    def set_source_expressions(self, exprs):
        self.y, self.x = exprs
```

coldmind replied:
I commented template and custom `as_sql`, and `tests.postgres_tests.test_aggregates.TestStatisticsAggregate#test_regr_avgx_with_related_obj_and_number_as_argument` failed, because X and Y was in wrong order.
I think we should explicitly see what is going on while constructing sql.
What do you think?

coldmind replied:
Oh, it was because I passed `self.x, self.y` instead of `self.y, self.x` to `parse_expressions`.
But question about explicitly is open.

jarshwah commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -0,0 +1,252 @@
+
+@unittest.skipUnless(connection.vendor == 'postgresql', 'PostgreSQL required')
+class TestGeneralAggregate(TestCase):
+    fixtures = ["aggregation_general.json"]
```
I think we should avoid writing new test suites that use fixtures. Fixture loading is extremely slow, and it's actually harder, IMO, to follow what the data should look like once you've aggregated it.

I would suggest either creating all the data in a setUp method, or creating the data you need at the top of each test.

charettes replied:
I suggest you use [setUpTestData](https://docs.djangoproject.com/en/dev/topics/testing/tools/#django.test.TestCase.setUpTestData) instead of `setUp`.

MarkusH replied:
:+1: on @charettes (thanks for the hint, didn't know this function exists :smile: )

coldmind commented:
@jarshwah anything else can be improved?
(and we didn't fully discussed the default alias for statistics functions)

jarshwah commented:
> Maybe we should give an ability users to override default alias in the way from your example: .aggregate(regr=RegrSYY('col_a', 'col_b'))

This is already possible. By using kwargs instead of args, the user supplies their own alias.

Regarding length, point taken. But I still don't think that this is a useful feature.

```
result = qs.aggregate(RegrSYY('hold_time', 'talk_time'))
result['hold_time_talk_time__regrsyy'] # ugh!
```

No existing aggregates take 2 input parameters, so it wasn't really a concern creating default aliases then. Also, the way it's implemented here, there are opportunities for clashing names:

```
.aggregate(RegrSYY(F('hold_time'), F('talk_time')) # num_num_regrsyy
.aggregate(RegrSYY(F('talk_time'), F('hold_time')) # num_num_regrsyy clash
```

My preference would be to deprecate default alias altogether, but that'd break a lot of older code for no real gain. We can prevent the spread of default alias though, which is why I'm discouraging it here.

I'd like input from others though. Is this a useful feature for these new aggregates as implemented? Or will it cause more troubles than it's worth?

jarshwah commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,103 @@
+
+    def __init__(self, y, x, output_field=FloatField()):
+        if not x or not y:
+            raise TypeError('Both X and Y must be provided.')
```
I think this should be a `ValueError`

jarshwah commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,103 @@
+]
+
+
+class StatFunc(Aggregate):
```
Perhaps a rename to StatAggregate?

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,216 @@
+``x`` and ``y``
+~~~~~~~~~~~~~~~
+
+Both of ``x`` and ``y`` can be a string with field name or integer.
```
or an expression returning an integer

charettes commented:
> My preference would be to deprecate default alias altogether, but that'd break a lot of older code for no real gain. We can prevent the spread of default alias though, which is why I'm discouraging it here.

I've always preferred to declare my alias explicitly so I wouldn't be sad to see the default one go away but I agree there's no real gain in deprecating it. +1 for not spreading it usage.

coldmind commented:
@jarshwah, @charettes I just removed default aliases.

mjtamlyn commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,197 @@
+.. warning::
+
+    All functions come without default alias,
+    so you must explicitly provide it.
```
This note needs more explanation I think - an example should suffice. Also it should be `aliases`.

mjtamlyn commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,197 @@
+
+.. class:: BitAnd(expression, **extra)
+
+    Returns the bitwise AND of all non-null input values, or null if none.
```
``AND``, ``NULL``

mjtamlyn commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,197 @@
+
+.. class:: BitOr(expression, **extra)
+
+    Returns the bitwise OR of all non-null input values, or null if none.
```
`OR`, `NULL`

mjtamlyn commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,197 @@
+``x`` and ``y``
+~~~~~~~~~~~~~~~
+
+Both of ``x`` and ``y`` can be a string with field name or an expression
```
Better:
> These aggregates take two required parameters - `x` and `y`. Each can either be a string representing a field name, or another expression returning an integer.

Does the field also need to represent an integer?

mjtamlyn commented:
It would be very beneficial to have examples and some more explanation of what the aggregates actually do in the documentation.

I don't know how much we want to explain the statistical functions - we have as much explanation as the postgres docs do. I don't think an understanding of regression analysis is "common knowledge" though, perhaps a link to the [wikipedia page](http://en.wikipedia.org/wiki/Regression_analysis) might be a good idea?

coldmind commented:
@mjtamlyn what examples do you want to see in the documentation?
I think we don't need to duplicate postgres documentation in that case.

mjtamlyn commented:
Some simple usage examples, they don't need to necessary show off the maths or the aggregation functionality, just how to use them. Perhaps not every one needs an example, but some would be helpful as it can make it much clearer what the aggregate does.

jarshwah commented:
The only other concern I have with this patch is the use of `x` and `y`. The functions expect Y first then X, which is consistent with the naming in the postgres docs. Traditionally X comes before Y though. The inconsistency is more apparent because the docs in this patch talk about "X and Y":

> Both of ``x`` and ``y`` can be a string with field name or an expression

If we can't (or shouldn't) rename the parameters from "y" and "x", we should at least be consistent in the documentation and talk about the parameters in the right order:

> Both of ``y`` and ``x`` can be a string with field name or an expression
> Both Y and X must be provided.

coldmind commented:
@jarshwah I'd prefer to use `Y` before `X`, because in postgres docs it defined in that way, and users will have ability to not explicitly specify arguments and use it like `SomeFunc('field1', 'field2')`.
I fixed documentation, thanks for remark.

@mjtamlyn I added some usage examples.

coldmind commented:
Any chance to move forward on this PR?

jarshwah commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -0,0 +1,41 @@
+    function = 'STRING_AGG'
+    template = "%(function)s(%(expressions)s, '%(delimiter)s')"
+
+    def __init__(self, *expressions, **extra):
```
I think the definition for this might be clearer as:

```
def __init__(self, expression, delimiter, **extra):
    super(StringAgg, self).__init__(expression, delimiter=delimiter, **extra)
```

The docs already have this definition, and multiple expressions do not make sense for this aggregate. You can also ditch the kwarg processing to check for delimiter (which is inconsistent with the docs at this point).

jarshwah commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,80 @@
+class StatAggregate(Aggregate):
+    def __init__(self, y, x, output_field=FloatField()):
+        if not x or not y:
+            raise ValueError('Both X and Y must be provided.')
```
"Both Y and X must be provided".

Switch the Y and X in the error.

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+.. warning::
+
+    All functions come without default aliases,
+    so you must explicitly provide it.
```
"provide one".

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+StringAgg
+~~~~~~~~~
+
+.. class:: String(expression, delimiter)
```
String -> StringAgg

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+
+    .. attribute:: delimiter
+
+        Required argument. Need to be a string. Examples: ``:``, ``;``.
```
Need**s** to be a string

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+``y`` and ``x``
+~~~~~~~~~~~~~~~
+
+Both of ``y`` and ``x`` can be a string with field name or an expression
```
"can be the name of a field or an expression"

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+~~~~~~~~~~~~~~~
+
+Both of ``y`` and ``x`` can be a string with field name or an expression
+returning a numeric data, and both must to be provided.
```
"returning numeric data. Both are required."

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+
+.. class:: RegrAvgX(y, x)
+
+    Returns average of the independent variable (``sum(x)/N``).
```
Returns the average

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+
+    .. attribute:: sample
+
+        By default, ``CovarPop`` returns the general population covariance. However,
```
remove the comma after default

jarshwah commented:
Other than the signature for StringAgg the code looks good to me. Someone will need to take a closer look at Docs though. 1.8beta is due for release very soon, so I'm fairly sure a lot of attention is focused there at the moment. You'll get some more traction once that's out the door.

coldmind commented:
@timgraham, can you take look at the docs?

coldmind commented:
Rebased on latest master, stilll waits for detailed review

akaariai commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,80 @@
+        if not x or not y:
+            raise ValueError('Both Y and X must be provided.')
+        super(StatAggregate, self).__init__(y=y, x=x, output_field=output_field)
+        self.x = x
```
Would self.set_source_expressions(self._parse_expressions(y, x)) work?

coldmind replied:
What is purpose of that?

akaariai commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,80 @@
+
+
+class StatAggregate(Aggregate):
+    def __init__(self, y, x, output_field=FloatField()):
```
Why y, x instead of x, y? Is this from PostgreSQL's function definition. If so, OK, if not, then I think the order should be x, y.

coldmind replied:
We discussed it here. Yes, postgresql use this order, so we should also use it in that way.

akaariai commented on django/contrib/postgres/aggregates/general.py:
```diff
@@ -0,0 +1,38 @@
+    template = "%(function)s(%(expressions)s, '%(delimiter)s')"
+
+    def __init__(self, expression, delimiter, **extra):
+        super(StringAgg, self).__init__(expression, delimiter=delimiter, **extra)
```
Is it sure StringAgg will return empty string instead of None always?

timgraham replied:
Do we have a test for this case?

coldmind replied:
@timgraham see https://github.com/django/django/pull/4105/files#r26198506

akaariai commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+   :synopsis: PosgreSQL specific aggregation functions
+
+All of these functions are available
+from the ``django.contrib.postgres.aggregates``.
```
In my opinion we should link to PostgreSQL documentation.

timgraham replied:
http://www.postgresql.org/docs/current/static/functions-aggregate.html

akaariai commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -0,0 +1,266 @@
+    def test_string_agg_empty_result(self):
+        GeneralTestModel.objects.all().delete()
+        values = GeneralTestModel.objects.aggregate(stringagg=StringAgg('char_field', delimiter=';'))
+        self.assertEqual(values, {'stringagg': None})
```
Hmmh, now I'm not sure if this is the right result. It seems array_agg will always return an array, but the same isn't true for string_agg. What is the rationale for this choice?

coldmind replied:
@akaariai, my opinion is that `ArrayAgg` is a special case, where we need to always return an array even if array will be empty.
So, maybe `StringAgg` should always return a string?

timgraham commented on django/contrib/postgres/aggregates/statistics.py:
```diff
@@ -0,0 +1,80 @@
+class StatAggregate(Aggregate):
+    def __init__(self, y, x, output_field=FloatField()):
+        if not x or not y:
+            raise ValueError('Both Y and X must be provided.')
```
lowercase y/x to match param names?

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+
+.. warning::
+
+    All functions come without default aliases,
```
break line closer to 79 characters
I think warning is a bit strong for this. 'note' is probably fine

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+    so you must explicitly provide one.
+    Usage example::
+
+        >> result = SomeModel.objects.aggregate(arr=ArrayAgg('somefield'))
```
>>> instead of >>
we can show the result directly as done in other examples instead of printing it

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+        >> print(result)
+        {'arr': [0, 1, 2]}
+
+
```
one space between sections

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+
+.. module:: django.contrib.postgres.aggregates
+   :synopsis: PosgreSQL specific aggregation functions
+
```
.. versionadded:: 1.9
(and release notes)

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+
+.. class:: ArrayAgg(expression, **extra)
+
+    Returns input values, including nulls, concatenated into an array.
```
can we rephrase each of these to say things like "Returns a list of values .." and get ride of "Return type: ..."

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+
+.. class:: StringAgg(expression, delimiter)
+
+    Returns input values concatenated into a string, separated by delimiter.
```
```
separated by the ``delimeter`` string.
```

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+
+    .. attribute:: delimiter
+
+        Required argument. Needs to be a string. Examples: ``:``, ``;``.
```
Not sure this is really needed given the above explanation. Shouldn't need to give examples of strings.

timgraham commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,235 @@
+    >> TestModel.objects.aggregate(count=RegrCount(y='field3', x='field2'))
+    {'count': 2}
+    >> TestModel.objects.aggregate(
+    avgx=RegrAvgX(y='field3', x='field2'), avgy=RegrAvgY(y='field3', x='field2')
```
... for line continuation

timgraham commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -0,0 +1,266 @@
+        StatTestModel.objects.create(
+            int1=1,
+            int2=3,
+            related_field=GeneralTestModel.objects.create(integer_field=0)
```
include trailing comma so if more kwargs are added, we don't need to modify this line again

timgraham commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -0,0 +1,266 @@
+            related_field=GeneralTestModel.objects.create(integer_field=2)
+        )
+
+    # Tests for class base
```
base class (StatAggregate)

timgraham commented on tests/postgres_tests/test_aggregates.py:
```diff
@@ -0,0 +1,266 @@
+
+    def test_correct_source_expressions(self):
+        func = StatAggregate(x='test', y=13)
+        self.assertTrue(isinstance(func.source_expressions[0], Value))
```
assertIsInstance

coldmind commented:
@timgraham , fixed your comments

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,232 @@
+    >> TestModel.objects.aggregate(result=ArrayAgg('field1'))
+    {'result': ['foo', 'bar', 'test']}
+
+Next example is about usage of statistics aggregation functions.
```
The next example shows the usage of statistical aggregate functions.

jarshwah commented on docs/ref/contrib/postgres/aggregates.txt:
```diff
@@ -0,0 +1,232 @@
+    {'result': ['foo', 'bar', 'test']}
+
+Next example is about usage of statistics aggregation functions.
+Underlying math will be not described (you can read about this,
```
The underlying math..

jarshwah commented:
I just added two more documentation suggestions, but other than that I'm happy for this to be merged.

coldmind commented:
@jarshwah thanks! fixed your comments

timgraham commented:
buildbot, test on precise.

timgraham commented:
Could you incorporate these doc edits if they look okay? http://dpaste.com/0R8XMDB
I thought we might rename `GeneralTestModel` to something like `AggregateTestModel` or `AggTestModel`. Please mark the ticket RFC when those changes are done.

coldmind commented:
@timgraham, fixed last comments.

timgraham commented:
merged in e4cf8c8420634d6f2dc8ce873246256ce635972d, thanks!
